### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by lroellin**
Also solves https://github.com/dwrtc/sa/issues/28

### CODE REVIEW

File: docs/signaling.md [hunk 1]
```diff
@@ -16,3 +16,43 @@ As we're in a P2P network, it is possible to add bad data. However, we wanted to
...
+
+The [ClientService] class is meant as the one-stop starting point for all DHT operations. It's meant as a singleton that creates and bootstraps the TomP2P peer. All objects are created through methods of this class.
+
+### Futures
+
+When we decided to build a nicer API for TomP2P Futures, we split it up.
+
+In this layer, you can find the high-level [Future]s. These are a rework of the API that TomP2P gives.
+
+In the [util][ch.hsr.dsl.dwrtc.util] layer, you can find the extension classes that build right on top of TomP2P. Our [Future] classes build on these.
```
Comment (thde) [inline docs/signaling.md]: I'm also not sure if we want to write something like "you can find". Shouldn't it be from a third person perspective? e.g. [util][ch.hsr.dsl.dwrtc.util] contains the extension classes that build right on top of TomP2P an is used by [Future].

File: docs/signaling.md [hunk 2]
```diff
@@ -16,3 +16,43 @@ As we're in a P2P network, it is possible to add bad data. However, we wanted to
...
+
+### InternalClient/ExternalClient
+
+We decided to make a distinction between those.
```
Comment (thde) [inline docs/signaling.md]: I think we can remove this one aswell.

File: docs/signaling.md [hunk 3]
```diff
@@ -16,3 +16,43 @@ As we're in a P2P network, it is possible to add bad data. However, we wanted to
...
+
+An [InternalClient] is created when we receive a new WebSocket session. It is able to send and receive messages. An [ExternalClient] is created when an [InternalClient] wants to send messages to it. It mainly consists of the found peer address this client is found at. An [ExternalClient] can only receive messages. On the other peer, the messages are then routed to a corresponding [InternalClient].
```
Comment (thde) [inline docs/signaling.md]: Here it "we receive". I'd also change this to a less personal style e.g. An [InternalClient] is created when a new WebSocket session is opened (don't think opened is perfect, neither is created).

File: docs/signaling.md [hunk 4]
```diff
@@ -16,3 +16,43 @@ As we're in a P2P network, it is possible to add bad data. However, we wanted to
...
+
+Note that, of course, an [ExternalClient] can be on the same server. This just makes handling things a lot nicer, since you don't have to care about such distinctions.
```
Comment (thde) [inline docs/signaling.md]: Why not just: An [ExternalClient] can be on the same server.

File: docs/signaling.md [hunk 5]
```diff
@@ -16,3 +16,43 @@ As we're in a P2P network, it is possible to add bad data. However, we wanted to
...
+
+ 
+### Interfaces
+
+On all these classes, we defined an Interface first before we did the actual implementation. This allows us to, in theory, swap the TomP2P layer underlying it. The interface is meant to be agnostic to the P2P layer it stands on.
```
Comment (thde) [inline docs/signaling.md]: All classes are implementing an interface for ("auswechselbarkeit"). The interface is meant to be agnostic to the P2P layer it stands on.

File: docs/websocket.md [hunk 6]
```diff
@@ -1,3 +1,48 @@
 # Package ch.hsr.dsl.dwrtc.websocket
 
-UI layer. Handles sending and receiving messages from the WebSocket connections. 
+UI layer. Handles sending and receiving messages from the WebSocket connections.
+
+## Design Decisions
+
+### Message format
+
+We're re-using the [Message] format for our WebSocket messages.
```
Comment (thde) [inline docs/websocket.md]: Would also reformat this.

File: docs/websocket.md [hunk 7]
```diff
@@ -1,3 +1,48 @@
...
+
+### Few Components
+
+The [WebSocketHandler] is relatively small. It just consists of four components:
```
Comment (thde) [inline docs/websocket.md]: The [WebSocketHandler] consists of four components:

File: docs/websocket.md [hunk 8]
```diff
@@ -1,3 +1,48 @@
...
+
+* [WebSocketHandler.clients] is a map of session ID to [InternalClient]s
+* [WebSocketHandler.sessions] is map of session ID to WebSocket sessions
+* [WebSocketHandler.onReceiveMessageFromWebSocket] uses the session ID to get the [InternalClient]. This is then used to send a message to send a message through the P2P layer
```
Comment (thde) [inline docs/websocket.md]: to send a message to send a message

File: docs/websocket.md [hunk 9]
```diff
@@ -1,3 +1,48 @@
...
+* [WebSocketHandler.onReceiveMessageFromSignaling] uses the session ID to get the WebSocket session. This is then used to send the message to the specific WebSocket
+
+The rest of the class are just some handlers that register the specific WebSocket callbacks to those methods.
```
Comment (thde) [inline docs/websocket.md]: I'd just remove this.

File: docs/websocket.md [hunk 10]
```diff
@@ -1,3 +1,48 @@
...
+
+## API Doc
+
+This API is quite simple.
```
Comment (thde) [inline docs/websocket.md]: I'd remove this aswell :).

File: docs/websocket.md [hunk 11]
```diff
@@ -1,3 +1,48 @@
...
+
+When connected to the WebSocket you can send or receive the following message types:
```
Comment (thde) [inline docs/websocket.md]: you

File: docs/websocket.md [hunk 12]
```diff
@@ -1,3 +1,48 @@
...
+
+### Send
+
+* `SignalingMessage`. Send your black box signaling message to another peer.
```
Comment (thde) [inline docs/websocket.md]: your

File: docs/websocket.md [hunk 13]
```diff
@@ -1,3 +1,48 @@
...
+  * `recipientSessionId: String`. The recipient's session ID. Used for routing.
+  * `messageBody: String`. The free-form message body. To the transport layer, this is a black box.
+  
+### Receive
+
+You must handle all these incoming message types. They are distinguishable by their `type` field.
```
Comment (thde) [inline docs/websocket.md]: All these incoming message types have to be handled.

File: docs/signaling.md [hunk 14]
```diff
@@ -16,3 +16,43 @@ As we're in a P2P network, it is possible to add bad data. However, we wanted to
 
 * [SignalingMessage.recipientSessionId] and [SignalingMessage.senderSessionId] IDs are overwritten before they go to the TomP2P layer, since they could be faked by WebSocket clients
```
Comment (thde) [inline docs/signaling.md]: SignalingMessage.recipientSessionId] and [SignalingMessage.senderSessionId] IDs are overwritten before they go to the TomP2P layer. (I'd remove the rest)

File: docs/signaling.md [hunk 15]
```diff
@@ -16,3 +16,43 @@ As we're in a P2P network, it is possible to add bad data. However, we wanted to
...
 * [ClientService.removeClient] only accepts an [IInternalClient] (that you got from [ClientService.addClient]), so you cannot easily disconnect another user.
```
Comment (thde) [inline docs/signaling.md]: you

File: docs/websocket.md [hunk 16]
```diff
@@ -1,3 +1,44 @@
 # Package ch.hsr.dsl.dwrtc.websocket
 
-UI layer. Handles sending and receiving messages from the WebSocket connections. 
+UI layer. Handles sending and receiving messages from the WebSocket connections.
+
+## Design Decisions
+
+### Message format
+
+To keep things simple, the [Message] format is re-used.
```
Comment (thde) [inline docs/websocket.md]: Maybe just: The [Message] format is re-used.

File: docs/util.md [hunk 17]
```diff
@@ -18,13 +18,13 @@ Note, that all these classes employ TomP2P semantics. Think of these as a very t
 
 `FutureGet` extension classes: these work on all Get operations that TomP2P returns. Also includes all `BaseFuture` extensions.
```
Comment (thde) [inline docs/util.md]: `FutureGet` extension classes work on all Get operations that TomP2P returns. Also includes all `BaseFuture` extensions.

File: docs/util.md [hunk 18]
```diff
@@ -18,13 +18,13 @@ Note, that all these classes employ TomP2P semantics. Think of these as a very t
...
 
-These fire on `BaseFuture`'s `onComplete`! Therefore, you have to check for success or failure yourself
+These fire on `BaseFuture`'s `onComplete`! Therefore, the success status still needs to be checked.
```
Comment (thde) [inline docs/util.md]: They fire on `BaseFuture`'s `onComplete`. Therefore, the success status still needs to be checked.

File: docs/signaling.md [hunk 19]
```diff
@@ -2,17 +2,55 @@
...
 
-To connect to other clients, you [find clients][ClientService.findClient]. This will get you an [ExternalClient]. You can then use an InternalClient to [send messages][InternalClient.sendMessage] to this ExternalClient. (This will call the message listener on the other peer's InternalClient you're sending messages to.)
+To connect to other clients, [find clients][ClientService.findClient]. This will return an [ExternalClient]. An [InternalClient] can then be used to [send messages][InternalClient.sendMessage] to this ExternalClient. (This will call the message listener on the other peer's InternalClient).
 
 ## Design decisions
 
 ### Anti-cheating mechanisms
 
-As we're in a P2P network, it is possible to add bad data. However, we wanted to make this a bit harder, so we at least don't expose it on our API.
+As this system in a P2P network, it is possible to add bad data. However, the API makes this harder:
```
Comment (thde) [inline docs/signaling.md]: As I think about it. I'd change the title and remove this sentence. Let's just say that the stuff get's overwritten.

File: docs/util.md (hunk 18 above)
Comment (lroellin) [reply docs/util.md]: This is just the dot vs. the exclamation mark, right?

**Pull request merged**